### DISCUSSION TITLE
Codable support

### DISCUSSION THREAD

**Discussion by Strobocop** [JosephDuffy/Partial#128] [Ideas]
This is a great library and a super useful type...so thanks for your work on this! 

One thing that would make this even more useful would be Codable support. I could see using this on the server side for patch updates. I admittedly haven't done a ton of research on the implementation of Partial yet, but wanted to get the conversation going. Do you see this being possible to implement Codable support? Guessing it my require a custom decoder/encoder or use of reflection due to the dynamic nature of the API. Wouldn't mind pitching in on this, but not sure if I would be on a wild goose chase and hitting limitations with Codable.

**Comment by JosephDuffy** [comment]
Sorry for the delay replying here @Strobocop. This sort of thing is something that I've been thinking about since I started work on this.

I've never come up with a way to make this work, although if you have some ideas I'd love to hear them.

The best I can come up with is a way to apply a `Partial` to a `KeyedEncodingContainer`. This would require some sort of method to map a `CodingKey` to a `KeyPath`. I think this would require a few tight constraints, along the lines of:

```swift
protocol KeyPathCodingKeyMapper {
  associatedtype Wrapped
  func codingKeyForKeyPath<Value: Codable, Key: CodingKey>(_ keyPath: KeyPath<Wrapped, Codable>) -> Key?
}

// ... inside Partial
func encode<Key: CodingKey>(to container: KeyedEncodingContainer<Key>, keyPathCodingKeyMapper: KeyPathCodingKeyMapper) {
  // Here we would loop over the stored keys, but how could you know that `Value` is `Codable`?
}
```

This is where I get stuck.

I've thought about updating some of the internals to support applying a `Partial` instance to a `Wrapped` instance by storing key paths that are `WritableKeyPath` (in this case it could be for `KeyPath`s with a `Value` that's `Codable`) but when you look at the internals the values are stored as `[PartialKeyPath<Wrapped>: Any]`.

Note that the `Value` is not known and the value is stored as `Any`.  Storing for example `[KeyPath<Wrapped, Codable>: Any]` would not be possible so this is where I get stuck with this idea.

**Comment by randomeizer** [comment]
How do you guys see `Codable` support working in this context exactly? Is it just mirroring the final 'unwrapped' value, or the partial values?

It's hard to guarantee that every given KeyPath will be Codable, but I guess it could just encode those which are?

Either way, I had a thought on how to get `Value` from a `PartialKeyPath<Wrapped>` value, namely:

```lua

protocol EncodableKeyPath {
    var rootType: Any.Type { get }
    var valueType: Encodable.Type { get }
}

extension KeyPath: EncodableKeyPath where Value: Encodable {
    var rootType: Any.Type { Root.self }

    var valueType: Encodable.Type { Value.self }
}
```

Looping through the internal values, you could do this:

```lua
for (keyPath, value) in values.enumerated() {
    guard let encodableKeyPath = keyPath as? EncodableKeyPath,
          let encodableValue = value as? Encodable,
          let codingKey = mapper.codingKeyForKeyPath(encodableKeyPath)
    else {
        throw Error.uncodableKeyPath(keyPath)
    }
    // encode the value...
    try container.encode(encodableValue, forKey: codingKey)
}
```

How would you see the `KeyPathCodingKeyMapper` getting initialised/passed around?

**Comment by JosephDuffy** [comment]
I played around with this again recently and feel like I have some ideas but nothing that works yet.

Conforming `Partial` to `Encodable`/`Decodable` seems to be impossible. The `KeyPath`/`CodingKey` mapper can't map for a `KeyPath` to a `CodingKey` because the return type (something conforming to `CodingKey`) can't be known. This may have been the easiest to implement, but also the least useful so I'm not too concerned about this.

I think the only possible solution to this is to create new coders, i.e. `PartialEncoder` and `PartialDecoder`. I've never done this before though.

From the research I've done it looks like it'll be easiest to have the new coders use another coder e.g. `JSONEncoder` for the "heavy lifting" of producing and consuming `Data`.

For decoding the values a mapper should be enough if it provides:

```swift
func keyPath<CodingKey: Swift.CodingKey>(for codingKey: CodingKey) -> PartialKeyPath<Self>?
```

I've not got any ideas for how to make encoding work (I don't know how to get the types working for `let codingKey = mapper.codingKeyForKeyPath(encodableKeyPath)`).

I will revisit this with a fresh mind again and see how far I can get.

**Comment by randomeizer** [comment]
Yeah, I had a play too, but couldn't come up with anything. Looked into using `Mirror` to work through available properties on the `Wrapped` type, but it's missing type information there too.

**Comment by JosephDuffy** [comment]
> How do you guys see `Codable` support working in this context exactly? Is it just mirroring the final 'unwrapped' value, or the partial values?

For this there are a couple of different ways it could be useful (if actually possible):

- Encode/decode a `Partial` instance
  - This enables a new feature: storing `Partial` instances e.g. for during a signup flow
- Decode a value of type `Wrapped` using a `Partial` instance
  - This would be better alternative to `PartialConvertible` 

I don't think encoding a value of type `Wrapped` into a `Partial` is possible or useful.

At one point I created an `init(wrapping:)` initialiser that would allow an instance to fallback to and existing value but I never added it because I thought it might be confusing and would be better suited as a separate `Mutable` type.

**Comment by JosephDuffy** [comment]
I've moved this to a discussion since it's available for this repo now.

I will keep this open and possibly revisit it if some new language features make this more feasible.

No one has a proof of concept for this but I am very open to new ideas or PRs.

I revisit this every few weeks thinking I've cracked it but I never come up with a solution that works!

**Comment by JosephDuffy** [comment]
I have a draft PR up with `Encodable` support: https://github.com/JosephDuffy/Partial/pull/175

I'd love your thoughts/feedback @Strobocop and @randomeizer!

**Reply by randomeizer** [reply]
Nice! I'll try have a play with it later today.

I think this is where my `EncodableKeyPath` concept above would come in helpful - it is essentially "exposing" the generic types of the`KeyPath` types, so that they can be used to cast outside. I'll see if I can whip up an implementation later.

Also, they can be totally private in the API, so no new public classes to worry about.

**Reply by randomeizer** [reply]
So I did have a play, but didn't really make any breakthroughs. The key challenges are linking KeyPath names to CodeKey values, which I don't think can be done automatically even by reflection currently. `EncodableKeyPath` _may_ be useful for type identification, but I didn't find a solution using it yet. Will let you know if I do...

**Reply by JosephDuffy** [reply]
Thanks for taking a look. I'm going to try and work on this some more (documentation, tests, etc.) but will wait a little bit to see if the unlocking of existential types gets through to see so how much that helps.

`@resultBuilder` was also added in Swift 5.4 (replacing `@_functionBuilder`) so perhaps there's something around that could help.

**Reply by Strobocop** [reply]
@JosephDuffy Sorry I have been zero help on this issue I originally created! Life and work have managed to crush my dreams of finding time to contribute to cool open source projects such as this. Glad to hear you have made some progress. Very interested to see if/how you pulled this off! Going to try to set aside some time to kick the tires.

Another development that may eventually be applicable (if it makes its way through the evolution process) is something the maintainers of the tensorflow branch of swift are working on:

https://github.com/tensorflow/swift/blob/main/docs/DynamicPropertyIteration.md
https://forums.swift.org/t/storedpropertyiterable/19218

**Reply by randomeizer** [reply]
Oh yeah, that would help a lot! Doesn't look like it has progressed much since 2019 though...

**Reply by JosephDuffy** [reply]
I think dynamic property iteration could help a lot. I've been hoping for it for a while so I can improve https://github.com/JosephDuffy/HashableByKeyPath. As I was thinking about it though I realised that `@resultBuilder`s (previously `@_functionBuilder`) could help in a similar way to how I've used it in https://github.com/JosephDuffy/HashableByKeyPath/pull/1.

I've now got the requirements down to a single function that provides a map between key paths and coding keys. With the current state of the language I don't think the API can be improved any further: https://github.com/JosephDuffy/Partial/pull/175/files#diff-cabcfef47af78d3bd5697f599ca40643a89b5feb1ee5e87fb7e53c01544cb5edR11

I plan to continue forward with this as the API, let me know if you can see anywhere else to improve it!

**Reply by randomeizer** [reply]
I added some comments on the current implementation. Personally, I'd be in favour of some variation of the `switch` method I proposed for linking `KeyPath` and `CodingKey` because it is less likely to result in missed keys when APIs change in the `Wrapped` type. I'm also not sure if you can make the @resultBuilder cached, or if it is basically regenerating the mapping every time it's called.
